src/anniversary/sf33rd/AcrSDK/ps2/flps2dma.c

```diff
@@ -1,111 +1,670 @@
+#include <eekernel.h>
+#include <eeregs.h>
+#include <libdma.h>
 
-INCLUDE_ASM("asm/anniversary/nonmatchings/sf33rd/AcrSDK/ps2/flps2dma", flPS2DmaAddRefTag);
+static __int128 vif1_fifo = (__int128)0x06000000;
+static u32 *dma_chcr_adrs[10] = { (u32 *)0x10008000, (u32 *)0x10009000, (u32 *)0x1000A000, (u32 *)0x1000B000,
+                                  (u32 *)0x1000B400, (u32 *)0x1000C000, (u32 *)0x1000C400, (u32 *)0x1000C800,
+                                  (u32 *)0x1000D000, (u32 *)0x1000D400 };
 
-INCLUDE_ASM("asm/anniversary/nonmatchings/sf33rd/AcrSDK/ps2/flps2dma", flPS2DmaAddRefeTag);
+u32 flPS2DmaAddCntTag(u32 tag, s32 qwc, s32 irq, s32 /*unused*/) {
+    QWORD *wk_ptr = (QWORD *)tag;
+    wk_ptr->UI128 = 0;
+    wk_ptr->UI32[0] = qwc + 0x10000000;
 
-#if defined(TARGET_PS2)
-INCLUDE_ASM("asm/anniversary/nonmatchings/sf33rd/AcrSDK/ps2/flps2dma", flPS2DmaAddEndTag);
-#else
-u32 flPS2DmaAddEndTag(u32 tag, s32 qwc, s32 irq, s32 /* unused */) {
-    not_implemented(__func__);
+    if (irq == 1) {
+        wk_ptr->UI32[0] |= 0x80000000;
+    }
+
+    return tag + 0x10;
+}
+
+u32 flPS2DmaAddRefTag(u32 tag, s32 qwc, u32 data_adrs, s32 irq, s32 /*unused*/) {
+    QWORD *wk_ptr = (QWORD *)tag;
+    u32 addr = data_adrs & 0xFFFFFFF;
+    u32 spr_flag = 0;
+
+    if ((data_adrs & 0x70000000) == 0x70000000) {
+        spr_flag = 0x80000000;
+    }
+
+    wk_ptr->UI128 = 0;
+    wk_ptr->UI32[0] = qwc + 0x30000000;
+    wk_ptr->UI32[1] = addr | spr_flag;
+
+    if (irq == 1) {
+        wk_ptr->UI32[0] |= 0x80000000;
+    }
+
+    return tag + 0x10;
 }
-#endif
 
-INCLUDE_ASM("asm/anniversary/nonmatchings/sf33rd/AcrSDK/ps2/flps2dma", flPS2VIF1CodeAddDirectHL);
+u32 flPS2DmaAddRefeTag(u32 tag, s32 qwc, u32 data_adrs, s32 irq, s32 /*unused*/) {
+    QWORD *wk_ptr = (QWORD *)tag;
+    u32 addr = data_adrs & 0xFFFFFFF;
+    u32 spr_flag = 0;
+
+    if ((data_adrs & 0x70000000) == 0x70000000) {
+        spr_flag = 0x80000000;
+    }
+
+    wk_ptr->UI128 = 0;
+    wk_ptr->UI32[0] = qwc;
+    wk_ptr->UI32[1] = addr | spr_flag;
+
+    if (irq == 1) {
+        wk_ptr->UI32[0] |= 0x80000000;
+    }
+
+    return tag + 0x10;
+}
+
+u32 flPS2DmaAddEndTag(u32 tag, s32 qwc, s32 irq, s32 /*unused*/) {
+    QWORD *wk_ptr = (QWORD *)tag;
+
+    wk_ptr->UI128 = 0;
+    wk_ptr->UI32[0] = qwc + 0x70000000;
+
+    if (irq == 1) {
+        wk_ptr->UI32[0] |= 0x80000000;
+    }
+
+    return tag + 0x10;
+}
+
+u32 flPS2VIF1CodeAddDirectHL(u32 tag, s32 qwc) {
+    QWORD *pkt = (QWORD *)tag;
+
+    pkt->UI32[0] = 0;
+    pkt->UI32[1] = 0;
+    pkt->UI32[2] = 0;
+    pkt->UI32[3] = qwc | 0x51000000;
+
+    return tag + 0x10;
+}
 
-#if defined(TARGET_PS2)
-INCLUDE_ASM("asm/anniversary/nonmatchings/sf33rd/AcrSDK/ps2/flps2dma", flPS2VIF1CalcLoadImageSize);
-#else
 u32 flPS2VIF1CalcLoadImageSize(u32 size) {
-    not_implemented(__func__);
+    u32 count;
+    u32 data_size;
+
+    data_size = 0;
+    count = size / 0x70000;
+
+    if ((size % 458752) != 0) {
```

Comment:
Makes sense to use a hex here
```suggestion
    if ((size % 0x70000) != 0) {
```